QuizApp/DataLayer/API/NetworkService.swift

```diff
@@ -0,0 +1,142 @@
+//
+//  NetworkService.swift
+//  QuizApp
+//
+//  Created by Matej Butkovic on 14.05.2021..
+//
+
+import Foundation
+
+enum RequestError: Error {
+    case clientError
+    case serverError
+    case noData
+    case dataDecodingError
+}
+
+enum Result<Success, Failure> where Failure: Error {
+    case success(Success)
+    case failure(Failure)
+}
+
+
+class NetworkService: NetworkServiceProtocol {
+
+
+    func executeUrlRequestNoResponse<String: Decodable>(_ request: URLRequest, completionHandler:
+            @escaping (Result<String, RequestError>) -> Void) {
+
+        let task = URLSession.shared.dataTask(with: request, completionHandler: { data, response, error -> Void in
+
+            guard let httpResponse = response as? HTTPURLResponse
+                    else {
+                completionHandler(.failure(.serverError))
+                return
+            }
+            if(httpResponse.statusCode == 200) {
+                completionHandler(.success("Saved" as! String))
+            }
+
+        })
+
+        task.resume()
+    }
+
+
+    func executeUrlRequest<T: Decodable>(_ request: URLRequest, completionHandler:
+            @escaping (Result<T, RequestError>) -> Void) {
+
+        let task = URLSession.shared.dataTask(with: request, completionHandler: { data, response, error -> Void in
+
+            guard let value = try? JSONDecoder().decode(T.self, from: data!) else {
+                completionHandler(.failure(.dataDecodingError))
+                return
+            }
+            completionHandler(.success(value))
+
+        })
+
+        task.resume()
+    }
+
+
+    func login(email: String, password: String, view: LoginPresenterProtocol?) -> Void {
+        guard let url = URL(string: "https://iosquiz.herokuapp.com/api/session?username=\(email)&password=\(password)") else {
+            return
+        }
+
+        var request = URLRequest(url: url)
+        request.httpMethod = "POST"
+        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+        executeUrlRequest(request) { (result: Result<Login, RequestError>) in
+            switch result {
+            case .failure(_):
+                DispatchQueue.main.async {
+                    view?.loginFail()
```

Comment:
You could've used the delegate pattern in the similar way you used this `LoginPresenterProtocol`.
```swift
// ViewController
... 
presenter.login(username: username, password: password)
...

extension LoginViewController: LoginPresenterDelegate {

    func loginFailed() {
        // do UI stuff
    }

}
```

```swift
// Presenter
...
weak var delegate: LoginPresenterDelegate?
...

func login(username: String, password: String) {
    networkService.login(username: username, password: password) { [weak self] (result: Result<Login, RequestError>) in 
        guard let self = self else { return }

        switch result {
        case .failure(_):
            self.loginFail()
        case .success(let value):
            self.loginSuccess(value)
        }
    }
}

func loginFail() {
    DispatchQueue.main.async {
        delegate.loginFailed()
    }
}

func loginSuccess(value) {
    // set user to userDefaults

    DispatchQueue.main.async {
        router.<goToMainScreen>
    }
}
```
```swift
// NetworkService
...
func login(email: String, password: String, @escaping completionHandler: (Result<Login, RequestError>) -> Void) {
    // setup URL and request

    executeUrlRequest(request, completionHandler: completionHandler)

    // or you could call handler explicitly
    executeUrlRequest(request) { (result: Result<Login, RequestError>) in
        completionHandler(result)
    }
}
```

What you'd be doing here is passing logic from view controller or presenter to other layers without them knowing what they're doing.
Since LoginViewController implements delegate, the presenter doesn't know that its delegate is UIViewController, it just knows it can call function `loginFailed()` and the delegate will take care of the rest.
NetworkService will receive a completion handler, rather than an object, and it will just call it, it shouldn't know that it needs to execute the code on the main thread or which method of the protocol or class is used here. Using the closure, the network service just provides a way for its callers to do some work once the network call returned a response.